File: dart/luhn/lib/luhn.dart
class Luhn {
  bool valid(String input) {
    if (input.trim().length <= 1) {
      return false;
    }
    var sum = 0;
    var zero = '0'.codeUnitAt(0);
    var nine = '9'.codeUnitAt(0);
    var counter = 0;
    for (var digit in input.split('').reversed) {
      if (digit != ' ') {
        var code = digit.codeUnitAt(0);
        if (code < zero || code > nine) {
          return false;
        }
        if (counter % 2 == 1) {
          sum += digit == '9' ? 9 : (int.parse(digit) * 2) % 9;
        } else {
          sum += int.parse(digit);
        }
        counter++;
      }
    }
    return sum % 10 == 0;
  }
}

File: kotlin/luhn/src/main/kotlin/Luhn.kt
package Luhn

fun isValid(s: String): Boolean {
  if (s.trim().length == 1) {
    return false
  }
  var sum = 0
  var counter = 0

  for (i in s.lastIndex downTo 0) {
    if (s[i].isDigit()) {
      var digit: Int = s[i].toInt() - '0'.toInt()
      counter += 1
      if (counter % 2 == 0) {
        val doubled = digit * 2
        sum += if (doubled > 9) doubled - 9 else doubled
      } else {
        sum += digit
      }
    } else if (s[i] != ' ') {
      return false
    }
  }
  return sum % 10 == 0
}